### PR TITLE
Correction to be compliant with i-Code & SonarQube

### LINKED ISSUE #6: i-Code plugin must be able to import standard i-Code result file
### Expected behavior
This plugin must be abble to import any file which is compliant to the official i-Code result XSD ([analysisResultXsd.zip](https://github.com/lequal/sonar-icode-cnes-plugin/files/1833880/analysisResultXsd.zip)). For example: it must be able to import a file containing issues for multiple source files.

### Actual behavior
This plugin is not able to load all compliant result files: it can only load issues of a result file for only one source file.

### Steps to reproduce behavior
* Set following settings:
  * extension for result files: .res
  * source files' suffixes : \*.f,\*.f77,\*.f90,\*.sh,\*.F,\*.F77,\*.F90
  * results folder : .
* Create a quality profile containing all rules and set it as default
* Execute sonar scanner in APSQ-001: [APSQ-001.zip](https://github.com/lequal/sonar-icode-cnes-plugin/files/1833838/APSQ-001.zip)

### Detection version
* Plugin version: V0
* SonarQube version: 6.7.1

### LINKED ISSUE #7: Source and result file names must be independant
### Expected behavior
The i-Code result file must be named and placed independently from analyzed source files.

### Actual behavior
Currently, the result file must be named like the analyzed source file (+ result file extension) and the result flder must be place at the same tree-level.

### Steps to reproduce behavior
* Set following settings:
  * extension for result files: .res
  * source files' suffixes : \*.f,\*.f77,\*.f90,\*.sh,\*.F,\*.F77,\*.F90
  * results folder : .
* Create a quality profile containing all rules and set it as default
* Execute sonar scanner in APSQ-002: [APSQ-002.zip](https://github.com/lequal/sonar-icode-cnes-plugin/files/1833950/APSQ-002.zip)

### Detection version
* Plugin version: V0
* SonarQube version: 6.7.1

### LINKED ISSUE #14: Errored metrics
### Expected behavior
Metrics reported in the view must reflect what i-Code analysis report.

### Actual behavior
Only metrics on functions are reported and no metrics on global file are reused (anoying for script language like shell).

### Steps to reproduce behavior
* Set following settings:
  * extension for result files: .res
  * source files' suffixes : \*.f,\*.f77,\*.f90,\*.sh,\*.F,\*.F77,\*.F90
  * results folder : .
* Create a quality profile containing all rules and set it as default
* Execute sonar scanner in APSQ-005 : [APSQ-005.zip](https://github.com/lequal/sonar-icode-cnes-plugin/files/1834002/APSQ-005.zip)
* Go on metrics page
* Compare with i-Code results

### Detection version
* Plugin version: V0
* SonarQube version: 6.7.1

### LINKED ISSUE #15: Sensor must analyze all the project tree.
### Expected behavior
When analyzing a project, sensor must browse all files in sub-folders.

### Actual behavior
Files in subfolders are ignored.

### Steps to reproduce behavior
* Set following settings:
  * extension for result files: .res
  * source files' suffixes : \*.f,\*.f77,\*.f90,\*.sh,\*.F,\*.F77,\*.F90
  * results folder : .
* Create a quality profile containing all rules and set it as default
* Execute sonar scanner in APSQ-004: [APSQ-004.zip](https://github.com/lequal/sonar-icode-cnes-plugin/files/1836535/APSQ-004.zip)

### Detection version
* Plugin version: V0
* SonarQube version: 6.7.1

### LINKED ISSUE #22: Plugin should implement Shell, Fortran 77 & 90 languages
### Expected behavior
With the plugin we should be able to differenciate all the three languages.

### Actual behavior
Therre is only one language called i-Code.

### Steps to reproduce behavior
Install the plugin and go and see languages in quality profiles' page.

### Detection version
1.0.0-alpha

### PR DESCRIPTION
**Pull request by begarco**
### Fixes :
- Fix #6
- Fix #7
- Fix #14 
- Fix #15
- Fix #22

---
This change is [Reviewable](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20)

### CODE REVIEW

File: src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	private static final String REPO_KEY_SUFFIX = "-rules";
+
+	/** Path to xml file in resources tree (shell rules). **/
+	public static final String PATH_TO_SHELL_RULES_XML = "/rules/icode-shell-rules.xml";
+
+	/** Path to xml file in resources tree (fortran 77 rules). **/
+	public static final String PATH_TO_F77_RULES_XML = "/rules/icode-f77-rules.xml";
+
+	/** Path to xml file in resources tree (fortran 90 rules). **/
+	public static final String PATH_TO_F90_RULES_XML = "/rules/icode-f90-rules.xml";
```
Comment (begarco) [inline src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java]: Code Smell Code Smell: Refactor your code to get this URI from a customizable parameter. ([squid:S1075](https://sonarcloud.io/organizations/lequal/rules#rule_key=squid%3AS1075))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=fr.cnes.sonarqube.plugins%3Asonar-icode-plugin&pullRequest=20&issues=AWSnS0Oosk71k7lw6nRN&open=AWSnS0Oosk71k7lw6nRN)

[ISSUE_KEY]: AWSnS0Oosk71k7lw6nRN

File: src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+ *
+ * @author lequal
+ */
+public class ICodeRulesDefinition implements RulesDefinition {
+
+	/** Partial key for repository. **/
+	private static final String REPO_KEY_SUFFIX = "-rules";
+
+	/** Path to xml file in resources tree (shell rules). **/
+	public static final String PATH_TO_SHELL_RULES_XML = "/rules/icode-shell-rules.xml";
```
Comment (begarco) [inline src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java]: Code Smell Code Smell: Refactor your code to get this URI from a customizable parameter. ([squid:S1075](https://sonarcloud.io/organizations/lequal/rules#rule_key=squid%3AS1075))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=fr.cnes.sonarqube.plugins%3Asonar-icode-plugin&pullRequest=20&issues=AWSnS0Oosk71k7lw6nRL&open=AWSnS0Oosk71k7lw6nRL)

[ISSUE_KEY]: AWSnS0Oosk71k7lw6nRL

File: src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+	/** Path to xml file in resources tree (fortran 77 rules). **/
+	public static final String PATH_TO_F77_RULES_XML = "/rules/icode-f77-rules.xml";
```
Comment (begarco) [inline src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java]: Code Smell Code Smell: Refactor your code to get this URI from a customizable parameter. ([squid:S1075](https://sonarcloud.io/organizations/lequal/rules#rule_key=squid%3AS1075))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=fr.cnes.sonarqube.plugins%3Asonar-icode-plugin&pullRequest=20&issues=AWSnS0Oosk71k7lw6nRM&open=AWSnS0Oosk71k7lw6nRM)

[ISSUE_KEY]: AWSnS0Oosk71k7lw6nRM

**Comment by WaldoFR** [comment]
___
*[CONTRIBUTING.md, line 1 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbRjGv93eEVKzAiY4T:-LHbRjGv93eEVKzAiY4U:b-cnc52v) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/5dfea6b88e00810fd3ebc33d10d1634183e4c4d4/CONTRIBUTING.md#L1)):*
> ```Markdown
> # Contributing to sonar-icode-cnes-plugin
> ```

Minor, why do you remove the "cnes" ?

**Comment by WaldoFR** [comment]
___
*[pom.xml, line 9 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbRwY16UWbyxKWsal0:-LHbRwY2EZE8YfobchdS:bq6bmtp) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/pom.xml#L9)):*
> ```XML
>     <packaging>sonar-plugin</packaging>
>     <version>1.0.0</version>
>     <name>Sonar i-Code plugin</name>
> ```

Sonar i-Code CNES plugin

**Comment by WaldoFR** [comment]
___
*[sonar-icode-plugin.iml, line 1 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbSZbg38ubwtcG2EK5:-LHbSZbh9mTQ0ZjYNFQT:bvo9fa6) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/sonar-icode-plugin.iml#L1)):*
> ```XML
> <?xml version="1.0" encoding="UTF-8"?>
> ```

Minor, can't be this file ignored in `.gitignore` ?

**Comment by WaldoFR** [comment]
___
*[src/main/java/fr/cnes/sonar/plugins/icode/ICodePlugin.java, line 29 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbT3xYCz09oG2C4FNi:-LHbT3xYCz09oG2C4FNj:b-lit5o) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/src/main/java/fr/cnes/sonar/plugins/icode/ICodePlugin.java#L29)):*
> ```Java
>  * This class is the entry point for all extensions.
>  * 
>  * @author Cyrille FRANCOIS
> ```

FYI no @author required

**Comment by WaldoFR** [comment]
___
*[src/main/java/fr/cnes/sonar/plugins/icode/languages/ShellLanguage.java, line 32 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbUDsNAh9xsWX4XFUh:-LHbUDsOC0YjNuedBuYc:b-i5ubwd) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/src/main/java/fr/cnes/sonar/plugins/icode/languages/ShellLanguage.java#L32)):*
> ```Java
> 	 * Name of the language.
> 	 */
> 	public static final String NAME = "Shell";
> ```

Minor, why not using properties file ?

**Comment by WaldoFR** [comment]
___
*[src/main/java/fr/cnes/sonar/plugins/icode/measures/ICodeNestingMetric.java, line 46 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbUdbYBv1XXdNCnMjT:-LHbUdbYBv1XXdNCnMjU:bl6x4ms) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/src/main/java/fr/cnes/sonar/plugins/icode/measures/ICodeNestingMetric.java#L46)):*
> ```Java
>             "Nesting (Maximum)",
>             Metric.ValueType.INT)
>             .setDescription("Maximum number of nested statements' levels.")
> ```

Minor typo

**Comment by WaldoFR** [comment]
___
*[src/main/java/fr/cnes/sonar/plugins/icode/settings/ICodePluginProperties.java, line 47 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbVHE1DzheKVH8I1mt:-LHbVHE1DzheKVH8I1mu:bmqoasw) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/src/main/java/fr/cnes/sonar/plugins/icode/settings/ICodePluginProperties.java#L47)):*
> ```Java
>      * Default value for the code suffix property
>      **/
>     public static final String SHELL_SUFFIX_DEFAULT = ".sh";
> ```

.ksh also

**Comment by WaldoFR** [comment]
___
*[src/main/java/fr/cnes/sonar/plugins/icode/settings/ICodePluginProperties.java, line 47 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbVMN9DJEgmPKiIlMa:-LHbVMN9DJEgmPKiIlMb:bocufhn) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/src/main/java/fr/cnes/sonar/plugins/icode/settings/ICodePluginProperties.java#L47)):*
> ```Java
>      * Default value for the code suffix property
>      **/
>     public static final String SHELL_SUFFIX_DEFAULT = ".sh";
> ```

Working, kornshell files too, `.ksh` and bash ones `.bash`

**Comment by WaldoFR** [comment]
___
*[src/main/resources/rules/icode-f90-rules.xml, line 7 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbXcSXCIgiZ_ibGWUH:-LHbXcSXCIgiZ_ibGWUI:b-gsmeg9) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/src/main/resources/rules/icode-f90-rules.xml#L7)):*
> ```XML
> 		<name>COM.DATA.FloatCompare</name>
> 		<internalKey>*</internalKey>
> 		<description>Unreferenced rule</description>
> ```

Minor there is a violation decription in i-Code CNES also for the F90 Rules

**Comment by WaldoFR** [comment]
___
*[src/main/resources/rules/icode-f77-rules.xml, line 7 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbXuk9ENKrVTU1Wbf7:-LHbXuk9ENKrVTU1Wbf8:b-dkuvpl) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/src/main/resources/rules/icode-f77-rules.xml#L7)):*
> ```XML
> 		<name>COM.DATA.FloatCompare</name>
> 		<internalKey>*</internalKey>
> 		<description>Unreferenced rule</description>
> ```

Minor "It's not allowed to compare float variables with equality."

**Comment by WaldoFR** [comment]
___
*[src/main/java/fr/cnes/sonar/plugins/icode/settings/ICodePluginProperties.java, line 65 at r1](https://reviewable.io/reviews/lequal/sonar-icode-cnes-plugin/20#-LHbbHvZ7VME1bbcXKH4:-LHbbHvZ7VME1bbcXKH5:b-vhpwl9) ([raw file](https://github.com/lequal/sonar-icode-cnes-plugin/blob/f6c0d38376fa80ee64d2ed5ea2c424c1e935bdff/src/main/java/fr/cnes/sonar/plugins/icode/settings/ICodePluginProperties.java#L65)):*
> ```Java
>      * Default value for the code suffix property
>      **/
>     public static final String F77_SUFFIX_DEFAULT = ".f,.f77,.F,.F77";
> ```

Minor, also `.for`,`.fpp`, `.fpp`,`.ftn`, `.FOR`,`.FPP`,`.FTN`

File: src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java (hunk 2 above)
Comment (begarco) [inline src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java]: Code Smell Code Smell: Refactor your code to get this URI from a customizable parameter. ([squid:S1075](https://sonarcloud.io/organizations/lequal/rules#rule_key=squid%3AS1075))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=fr.cnes.sonarqube.plugins%3Asonar-icode-plugin&pullRequest=20&issues=AWSnS0Oosk71k7lw6nRL&open=AWSnS0Oosk71k7lw6nRL)

[ISSUE_KEY]: AWSnS0Oosk71k7lw6nRL

File: src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java (hunk 3 above)
Comment (begarco) [inline src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java]: Code Smell Code Smell: Refactor your code to get this URI from a customizable parameter. ([squid:S1075](https://sonarcloud.io/organizations/lequal/rules#rule_key=squid%3AS1075))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=fr.cnes.sonarqube.plugins%3Asonar-icode-plugin&pullRequest=20&issues=AWSnS0Oosk71k7lw6nRM&open=AWSnS0Oosk71k7lw6nRM)

[ISSUE_KEY]: AWSnS0Oosk71k7lw6nRM

File: src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java (hunk 1 above)
Comment (begarco) [inline src/main/java/fr/cnes/sonar/plugins/icode/rules/ICodeRulesDefinition.java]: Code Smell Code Smell: Refactor your code to get this URI from a customizable parameter. ([squid:S1075](https://sonarcloud.io/organizations/lequal/rules#rule_key=squid%3AS1075))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=fr.cnes.sonarqube.plugins%3Asonar-icode-plugin&pullRequest=20&issues=AWSnS0Oosk71k7lw6nRN&open=AWSnS0Oosk71k7lw6nRN)

[ISSUE_KEY]: AWSnS0Oosk71k7lw6nRN

**Pull request merged**